mycli/main.py

```diff
@@ -387,26 +380,63 @@ def connect(self, database='', user='', passwd='', host='', port='',
 
         # Connect to the database.
 
-        try:
+        def _connect():
             try:
-                sqlexecute = SQLExecute(database, user, passwd, host, port,
-                        socket, charset, local_infile, ssl)
+                self.sqlexecute = SQLExecute(database, user, passwd, host, port,
+                                             socket, charset, local_infile, ssl)
             except OperationalError as e:
                 if ('Access denied for user' in e.args[1]):
-                    passwd = click.prompt('Password', hide_input=True,
-                                          show_default=False, type=str)
-                    sqlexecute = SQLExecute(database, user, passwd, host, port,
-                            socket, charset, local_infile, ssl)
+                    new_passwd = click.prompt('Password', hide_input=True,
+                                              show_default=False, type=str)
+                    self.sqlexecute = SQLExecute(database, user, new_passwd, host, port,
+                                                 socket, charset, local_infile, ssl)
                 else:
                     raise e
+
+        try:
+            if socket is host is port is None:
```

Comment:
It's possible that these are blank, but not `None`, e.g. an empty string `''`.

Reply:
Empty strings are never valid values for any of these options (or for any other mysql options?), so maybe the loader functions should do that validation/coercion?

Otherwise

```python
if all((opt in (None, '') for opt in (socket, host, port))):
```

might be a reasonably readable way of rewriting it.

Reply:
https://github.com/keleshev/schema is a validation library that's pretty simple

Reply:
Why not `if not (socket or host or port)`?

Reply:
I don't want to treat `port = 0` as 'no value', though it is treated as 'no value' by the `port = port or 3306` line, now that I think about it, so I guess it doesn't matter - problems are just being moved along to pymysql.

I don't know how good pymysql's error messages are for this kind of bad config. If they're good enough maybe there's no point doing your own validation.

Reply:
@cmcaine True, I forgot that port may be 0. However, I just tried changing default port to 0 on my local mysql installation, and it looks like the service defaults to 3306 in this case. Here is my.cnf:
```
--- ~ » cat /etc/my.cnf
[client]
port = 0
socket = /tmp/mysql.sock

[mysqld]
port = 0
socket = /tmp/mysql.sock
key_buffer_size = 16M
max_allowed_packet = 8M

[mysqldump]
quick
```

but here is what mysql thinks the port is set to:

```
mysql root@localhost:test> show global variables like 'port%';
+-----------------+---------+
| Variable_name   | Value   |
|-----------------+---------|
| port            | 3306    |
+-----------------+---------+
```

I think we can assume that 0 is not a valid port. Also, I'd rather defer to pymysql and not add more complexity with our own validation.

Reply:
Port 0 is never a valid port in TCP or UDP/IP.

If you connect to localhost, `mysql` ignores any port value and always uses the socket. I don't know how it does this. Presumably by reading the mysqld parts of the config files. Maybe a default is baked into the build.

If you have more than one server running on localhost. I think you have to set host to 127.0.0.1 and port to \<whatever> or just set socket to db2's socket. Which is confusing: `mysql -P 3307` shouldn't silently transmute to `mysql -S /path/to/db1.sock`.

> but here is what mysql thinks the port is set to:

My understanding is that `show global variables like 'port%';` shows the server's port, not the method that the client is using to connect. Try `status` instead.

---

I think this algorithm is least surprising:

1. If port is set, use TCP/IP
2. If UNIX socket is set use that
3. If both are set, throw an error
4. If neither are set:
    1. use TCP/IP if hostname is set and not localhost or if system is windows
    2. otherwise use UNIX socket at a default location, then maybe fall back to localhost:3306.

I think specifying both port and socket should be an error and that the current behaviour is bad:

 - mycli ignores socket if host or port are set at the commandline, but ignores the port otherwise, 
 - pymysql ignores the port,
 - mysql always uses socket whether you ask it to or not for localhost, 
 - All seem to ignore socket if host is not localhost (I get [1698](https://dev.mysql.com/doc/refman/5.7/en/error-messages-server.html#error_er_access_denied_no_password_error) if user is only identified by unix_sockets for `host == <public IP>`.

Best behaviour if config and commandline disagree would be, IMHO, to emit a warning that the config value for socket is being dropped in favour of the commandline argument. Also acceptable to just throw an error and refuse to connect.

Reply:
The above sounds sensible. But I don't think that we should throw an error in 3. I think if port is set, we should just prefer that over socket.

Reply:
That's the opposite behaviour to mysql.

Reply:
Really? mysql will prefer socket if port is set in the command line, as in `mysql --port 5555`?

Reply:
@cmcaine Just checked - yes it will. That behavior does not seem very user-friendly though. I think the user would expect to be connected the way he specified when running the CLI, i.e. if port or host are specified, they should be the preferred method.

Reply:
I already did the checks and listed the behaviour of each tool, and said that it's bad :)

Commandline taking precedence over config is fine. I'd maybe emit a warning, but whatever. If port and socket are both set on the commandline, I think that should be an error. If they're both set in the config file, I think that should be an error.

I don't want to bikeshed this any more. This is a pretty small convenience feature. I've said what I'd do, but I don't really care one way or the another.

Someone with investment in this project should just take these comments and finish them up in a reasonably sane way (please not how mysql does it).

Reply:
I was not trying to argue with you @cmcaine - you did good work on the PR. The results of your testing the tools above are also very helpful. My comments were only suggestions.

Reply:
I didn't feel like you were, and I didn't intend to be rude. Your suggestions were helpful and have clarified the spec :)